core/deadline.go

```diff
@@ -197,37 +202,75 @@ func getCurrDuty(duties map[Duty]bool, deadlineFunc DeadlineFunc) (Duty, time.Ti
 	return currDuty, currDeadline
 }
 
-// NewDutyDeadlineFunc returns the function that provides duty deadlines or false if the duty never deadlines.
-func NewDutyDeadlineFunc(ctx context.Context, eth2Cl eth2wrap.Client) (DeadlineFunc, error) {
+// NewDutyDeadlineFunc returns two functions:
+// 1. dutyDeadline: Determines the deadline for a duty to be processed. If a duty never expires (e.g., exit or registration duties), it returns false.
+// 2. submitDutyDeadline: Determines the submission deadline for a duty, which varies based on the duty type.
+// For example, proposer duties have a submission deadline of 1/3 of the slot duration.
+// Both functions use the genesis time and slot duration to calculate deadlines.
+func NewDutyDeadlineFunc(ctx context.Context, eth2Cl eth2wrap.Client) (DeadlineFunc, SubmitDeadlineFunc, error) {
 	genesis, err := eth2Cl.Genesis(ctx, &eth2api.GenesisOpts{})
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	genesisTime := genesis.Data.GenesisTime
 
 	eth2Resp, err := eth2Cl.Spec(ctx, &eth2api.SpecOpts{})
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	duration, ok := eth2Resp.Data["SECONDS_PER_SLOT"].(time.Duration)
+	slotDuration, ok := eth2Resp.Data["SECONDS_PER_SLOT"].(time.Duration)
 	if !ok {
-		return nil, errors.New("fetch slot duration")
+		return nil, nil, errors.New("fetch slot duration")
+	}
+
+	delta := slotDuration * time.Duration(lateFactor)
+	if delta < lateMin {
+		delta = lateMin
 	}
 
-	return func(duty Duty) (time.Time, bool) {
+	dutyDeadline := func(duty Duty) (time.Time, bool) {
 		if duty.Type == DutyExit || duty.Type == DutyBuilderRegistration {
 			// Do not timeout exit or registration duties.
 			return time.Time{}, false
 		}
 
-		start := genesisTime.Add(duration * time.Duration(duty.Slot))
-		delta := duration * time.Duration(lateFactor)
-		if delta < lateMin {
-			delta = lateMin
-		}
+		start := genesisTime.Add(slotDuration * time.Duration(duty.Slot))
 		end := start.Add(delta)
 
 		return end, true
-	}, nil
+	}
+
+	submitDutyDeadline := func(duty Duty) (time.Time, bool) {
+		if duty.Type == DutyExit || duty.Type == DutyBuilderRegistration {
```

Comment:
I think to refactor this entirely: to make an object for each duty that is responsible for tracking its own deadlines. Having common mockable interface of course.